pkg/qrcodegenerator/qrcode/gateways/ttigpro1.go

```diff
@@ -0,0 +1,98 @@
+// Copyright © 2024 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gateways
+
+import (
+	"regexp"
+
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/v3/pkg/types"
+)
+
+const (
+	formatIDttigpro1 = "ttigpro1"
+
+	euiLength        = 16
+	ownerTokenLength = 12
```

Comment:
This is now declared at the package level, which shouldn't be. Can you move this in a package?

Reply:
I'm not sure which package would be appropriate for these constants 🤔 Or do you mean that the full format should be in a package called ttigpro1?

Reply:
Yes

Reply:
If the format is moved into its own package, there are a lot of changes that have to be made because of how the gateways.Server Parse method is implemented to support multiple formats (we have only one format for now, but I followed the same implementation approach that's used for the end devices in case we add more later). Parse checks the provided format and instantiates it:
```go
// pkg/qrcodegenerator/qrcode/gateways/gateways.go

f := gtwFormat.format.New()
if err := f.UnmarshalText(data); err != nil {
	return nil, err
}
```
Because of this, the format interface must have the `New() Data` method. But the return value `Data` is also an interface. If the format is moved to its own ttigpro1 package, that package must import Data from the gateways package to implement the `New() Data` method:
```go
// pkg/qrcodegenerator/qrcode/gateways/ttigpro1/ttigpro1.go

import "go.thethings.network/lorawan-stack/v3/pkg/qrcodegenerator/qrcode/gateways"

// New implements the Format interface.
func (Format) New() gateways.Data {
	return new(ttigpro1)
}
```

But the gateways package also imports the ttigpro1 format from the ttigpro1 package so it can initialise it in the gatewaysFormats:
```go
// pkg/qrcodegenerator/qrcode/gateways/gateways.go

import "go.thethings.network/lorawan-stack/v3/pkg/qrcodegenerator/qrcode/gateways/ttigpro1"

func New(_ context.Context) *Server {
	s := &Server{
		// Newer formats should be added to this slice first to
		// preferentially match with those first.
		gatewayFormats: []gatewayFormat{
			{
				id:     ttigpro1.FormatID,
				format: new(ttigpro1.Format),
			},
		},
	}
	return s
}
```

Thus resulting in an import cycle. I can think of another way of restructuring things if it's fine to be inconsistent with how the end device formats are implemented.

Reply:
In Go, you don't need to return the type. You can return a struct that implements the interface. The `ttigpro1` package would not need to import anything from `gateways`: there are no types in `Data` that aren't known to a sub package.

So if you move this under `ttigpro1` and rename the struct to `Data`, then you can just instantiate `&ttigpro1.Data{}` from the `gateways` package. Maybe I miss something, but I don't see anything here that causes an import cycle.

Reply:
Thing is, Data is not a struct, it's an interface defined in the gateways package:
```go
// pkg/qrcodegenerator/qrcode/gateways/gateways.go
// Data represents gateway QR code data.
type Data interface {
	// FormatID returns the ID of the format used to parse the QR Code data.
	FormatID() string
	GatewayEUI() types.EUI64
	OwnerToken() string
	encoding.TextUnmarshaler
}
```

The ttigpro1 type implements the Data interface. This type is not instantiated directly, it is instantiated by a wrapper type, TTIGPRO1Format. This wrapper type implements another interface, Format:

```go
// pkg/qrcodegenerator/qrcode/gateways/gateways.go
type Format interface {
	Format() *ttnpb.QRCodeFormat
	New() Data // Data is an interface
}
```

The New() method of the Format interface is called in the Parse() method of the gateways.Server based on the format ID which is provided in the grpc call (or it defaults to the first format) - it's implemented in this way to have the Parse() method independent of the specific formats:
```go
// pkg/qrcodegenerator/qrcode/gateways/gateways.go

// Parse the given QR code data. If formatID is provided, only that format is used.
// Otherwise, the first format registered will be used.
func (s *Server) Parse(formatID string, data []byte) (ret Data, err error) {
	for _, gtwFormat := range s.gatewayFormats {
		// If format ID is provided, use only that. Otherwise,
		// default to the first format listed in gatewayFormats.
		if formatID != "" && formatID != gtwFormat.id {
			continue
		}

		f := gtwFormat.format.New() // this is where ttigpro1 is instantiated (i.e. the Data interface), but through the New method of the Format interface
		if err := f.UnmarshalText(data); err != nil {
			return nil, err
		}

		return f, nil
	}

	return nil, errUnknownFormat
}
```

I followed this pattern because it's used in the end device formats too, and it makes it easy to add new formats in the future.

If I move the ttigpro1 type (which implements the Data interface and which now will be  private) to its own package, the New() method in the wrapper type for ttigpro1 which implements the Format interface will look like this:

```go
// pkg/qrcodegenerator/qrcode/gateways/ttigpro1/ttigpro1.go
// New implements the Format interface.
func (Format) New() *ttigpro1 { // *ttigpro1 implements the gateways.Data interface
	return new(ttigpro1)
}

```

And when I try to instantiate the ttigpro1.Format:

```go
// pkg/qrcodegenerator/qrcode/gateways/gateways.go

// New returns a new gateways.Server.
func New(_ context.Context) *Server {
	s := &Server{
		// Newer formats should be added to this slice first to
		// preferentially match with those first.
		gatewayFormats: []gatewayFormat{
			{
				id:     ttigpro1.Format,
				format: new(ttigpro1.Format), // this is where the error pops up
			},
		},
	}
	return s
}
```

This error happens, even though ttigpro1 implements the Data interface:  

<img width="1087" alt="Screenshot 2024-09-09 at 11 17 16" src="https://github.com/user-attachments/assets/7b73962a-df44-4084-ae6a-e205bba0fe14">    

If we want to move the ttigpro1 to its separate package, one solution would be to move the gateways.Data interface that is now defined in the gateways package to the qrcode package, which will become qrcode.GatewayData (as there is already a qrcode.Data interface for the end devices):

```go
// pkg/qrcodegenerator/qrcode/qrcode.go

// Data represents QR code data. -> this is for end devices
type Data interface {
	Validate() error
	encoding.TextMarshaler
	encoding.TextUnmarshaler
}

// GatewayData represents gateway QR code data.
type GatewayData interface {
	// FormatID returns the ID of the format used to parse the QR Code data.
	FormatID() string
	GatewayEUI() types.EUI64
	OwnerToken() string
	encoding.TextUnmarshaler
}
```

Then the import cycle issue will be solved. Would that work?

Reply:
OK don't bother then and make sure symbols defined in `ttigpro1.go` are specific to TTIGPRO1.

Reply:
Note that `euiLength` doesn't seem to be used now, and `ownerTokenLength` has one usage that you might as well just inline.

Reply:
inlined